File: leetcode/linked_list_cycle.js
/* Given a linked list, determine if it has a cycle in it.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    for (var slow = head, fast = head; fast; ) {
        slow = slow.next;
        fast = fast.next;
        if (!fast)
            break;
        fast = fast.next;
        if (slow === fast)
            return true;
    }
    return false;
};

File: lintcode/linked_list_cycle.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: True if it has a cycle, or false
    """
    def hasCycle(self, head):
        slow = fast = head
        while fast:
            slow, fast = slow.next, fast.next
            if not fast: break
            fast = fast.next
            if slow == fast: return True
        return False
